feature: add helm push

h314to opened the pull request:
Typically most of the default options can be left unchanged, and the step will be just:

```
      - codacy/helm_push:
          name: helm_push_to_integration
          context: CodacyAWS
          chart_name: portal
          aws_profile: integration
```

andreaTP commented on src/jobs/helm_push.yaml:
```diff
@@ -0,0 +1,137 @@
+    type: string
+    description: "aws region for the deployment"
+    default: eu-west-1
+  credentials_file_content:
```
this entire section is already in: https://github.com/codacy/codacy-orbs/blob/master/src/jobs/aws_deploy.yml#L17
can  we refactor in a way that stays only once in the codebase?
e.g.: build a configurable command and re-use it around

h314to replied:
We can have a step that creates credentials file and persist it to the workspace for later commands to use, like we do with .version. If you think that is ok I can take care of it.

andreaTP commented on src/jobs/helm_push.yaml:
```diff
@@ -0,0 +1,137 @@
+        CHART_FILE="${HELM_DIR}/${CHART_NAME}/Chart.yaml"
+        VALUES_FILE="${HELM_DIR}/${CHART_NAME}/values.yaml"
+
+        if [ -z $APP_VERSION ] ; then
```
instead of binding parameters:
from user -> to env vars -> to bash scripts

please use orbs conditionals and avoid unuseful re-bind
e.g.:
https://github.com/codacy/codacy-orbs/blob/master/src/jobs/checkout_and_version.yml#L23

h314to replied:
It is a good idea. I'll look into it.

paulopontesm commented on src/commands/run_aws.yml:
```diff
@@ -0,0 +1,51 @@
+      role_arn = arn:aws:iam::$AWS_ACCOUNT_ID_PRODUCTION:role/$CONTINUOUS_DELIVERY_ROLE
+  deploy:
+    type: boolean
+    description: "Use deploy instead of run for this command"
```
```suggestion
    description: "Use 'deploy' instead of 'run' for this command"
```

h314to replied:
Yes, good idea.

paulopontesm commented on .circleci/config.yml:
```diff
@@ -1,120 +1,108 @@
-  orb-tools: circleci/orb-tools@8.5.0
+  codacy: codacy/base@0.0.13
+
+references:
```
do you need to deliver the changes for this file in this PR?

h314to replied:
Nope, this was just based on top of the circleci branch. Now that it is merged I'll rebase on master and these changes will disappear.

h314to commented:
I rebased on the current master, and implemented the following requested changes:

- refactored the aws specific setup into a command.
- added a tool to set values in yaml files instead of using sed
- renamed the 'helm' job to 'helm_aws' to denote that it does aws specific stuff (like setting up the cluster)
- improved usage of circleci parameters, and simplified job scripts 
- added examples

andreaTP commented on src/jobs/aws_deploy.yml:
```diff
@@ -37,13 +22,8 @@ environment:
+
+  - run_aws:
+      title: AWS deploy command - << parameters.cmd >>
+      deploy: true
```
I linke this! but be sure we are consistent in choosing defaults

andreaTP commented on src/jobs/helm_aws.yml:
```diff
@@ -0,0 +1,53 @@
+      steps:
+        - run_aws:
+            title: Setup kubeconfig context for the cluster
+            cmd: aws eks update-kubeconfig --name << parameters.cluster_name >>
```
since this command is executed "under the hood" I think is useful to increase the logging level,
e.g. `--debug`

h314to replied:
It's a good idea, but it would leak credentials to the log. I'll show contexts after updating kubeconfig to increase visibility about what this step does.

andreaTP commented on src/jobs/helm_push.yaml:
```diff
@@ -0,0 +1,39 @@
+  - run:
+      name: Set image tag and version using semver
+      command: |
+        ytool -s appVersion "$(cat .version)" -s version "$(cat .version)" -f "${HELM_DIR}/${CHART_NAME}/Chart.yaml" -e
```
instead of mangling the yml files can we enforce some good practice on the Helm files themselves to inject these vaules, or alternatively to use/build a simple helm plugin to handle this requirement?

paulopontesm replied:
What is the problem with this approach?

paulopontesm replied:
Right now we just want a way to bump versions of an helm chart before release. This approach solves our problem by treating the file for what it is.. a yml file.

In my point of view a simple `sed` in here would do it just fine, but I appreciate the extra-mile of having a tool that does this specific for yaml (without loosing the structure and comments).

andreaTP replied:
The problem is that we are mangling source code in CI before pushing a stable artifact, in this way we loose the source of truth.

Can we inject the version by leveraging Jinja templates?
I appreciate the effort of not using a plain `sed` but I'm confident we have other approaches to the problem itself.
cc. @rtfpessoa

paulopontesm replied:
Helm dosen't use jinja templates. We would have to add a new tool for that, this tool would be more complicated than the one we are using here. Also it wouldn't solve the problem you stated:

>  mangling source code in CI before pushing a stable artifact, in this way we loose the source of truth. 

There is no way of having the source of truth for the version in the repo without commiting and push.

Do you have a proposal to solve this problem?

andreaTP replied:
in the other repository we avoid to store the version in the files at all, and we inject in CI, whatever method you find to do it is fine to me.
